src/mongoose_commands.erl

```diff
@@ -344,21 +348,23 @@ unregister_commands(Commands) ->
         end,
         Commands).
 
+-spec execute_command(caller(), atom() | t(), args()) ->
+    success() | failure().
 execute_command(Caller, Command, Args) ->
     try check_and_execute(Caller, Command, Args) of
         ignore_result ->
             ok;
-        Res ->
+        {error, Type, Reason} ->
+            {error, Type, Reason};
+        {ok, Res} ->
             {ok, Res}
     catch
-        {badarg, E} ->
-            {error, type_error, E};
-        {type_error, E} ->
-            {error, type_error, E};
-        permission_denied ->
-            {error, denied, <<"Command not available for this user">>};
-        caller_jid_mismatch ->
-            {error, denied, <<"Caller ids do not match">>};
+        % admittedly, not the best style of coding, in Erlang at least. But we have to do plenty
+        % of various checks, and in absence of something like Elixir's "with" construct we are
+        % facing a choice between throwing stuff or using some more or less tortured syntax
+        % to chain these checks.
+        throw:{Type, Reason} ->
+            {error, Type, Reason};
```

Comment:
🤯 
If we want to run a list of checks and stop on an error, instead of throwing to escape the function flow, I'd do a fold over a list of functions that would stop when an error is detected, which is a monad style. Monads are not really idiomatic of Erlang, but more and more people do them in the functional world instead of the old throw-trick (and obviously better than a big bunch of nested case expressions xD). Anyway it might be out of the scope of this ticket, but in case we want to clean this code even more, it's an idea.

Reply:
yes, and better than chaining `maybe_*` functions :)